system/lib/libc/dynlink.c

```diff
@@ -50,15 +50,19 @@ static struct dso * _Atomic head, * _Atomic tail;
 
 #ifdef _REENTRANT
 static thread_local struct dso* thread_local_tail;
-static pthread_rwlock_t lock;
+static pthread_mutex_t write_lock = PTHREAD_MUTEX_INITIALIZER;
 
-static void dlsync_locked() {
+static void dlsync() {
   if (!thread_local_tail) {
     thread_local_tail = head;
   }
+  if (!thread_local_tail->next) {
+    return;
+  }
+  dbg("dlsync: catching up %p %p", thread_local_tail, tail);
   while (thread_local_tail->next) {
```

Comment:
Why is it ok to loop in this linked list without taking a lock? Can't it be modified by dlopen at the same time?

Reply:
This is a linked list that only grows and is never mutates, so it can be traversed in a lock-free way like this.